homeassistant/components/zwave/migration.py

```diff
@@ -0,0 +1,166 @@
+"""Handle migration from legacy Z-Wave to OpenZWave and Z-Wave JS."""
+from __future__ import annotations
+
+from collections import defaultdict
+from typing import TYPE_CHECKING, TypedDict, cast
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import async_get as async_get_device_registry
+from homeassistant.helpers.entity_registry import async_get as async_get_entity_registry
+from homeassistant.helpers.singleton import singleton
+from homeassistant.helpers.storage import Store
+
+from .const import DOMAIN
+from .util import node_device_id_and_name
+
+if TYPE_CHECKING:
+    from . import ZWaveDeviceEntityValues
+
+LEGACY_ZWAVE_MIGRATION = f"{DOMAIN}_legacy_zwave_migration"
+STORAGE_WRITE_DELAY = 30
+STORAGE_KEY = f"{DOMAIN}.legacy_zwave_migration"
+STORAGE_VERSION = 1
+
+
+class ZWaveMigrationData(TypedDict):
+    """Represent the Z-Wave migration data dict."""
+
+    node_id: int
+    node_instance: int
+    command_class: int
+    command_class_label: str
+    value_index: int
+    device_id: str
+    domain: str
+    entity_id: str
+    unique_id: str
+    unit_of_measurement: str | None
+
+
+@callback
+def async_is_ozw_migrated(hass):
+    """Return True if migration to ozw is done."""
+    ozw_config_entries = hass.config_entries.async_entries("ozw")
+    if not ozw_config_entries:
+        return False
+
+    ozw_config_entry = ozw_config_entries[0]  # only one ozw entry is allowed
+    migrated = bool(ozw_config_entry.data.get("migrated"))
+    return migrated
+
+
+@callback
+def async_is_zwave_js_migrated(hass):
+    """Return True if migration to Z-Wave JS is done."""
+    zwave_js_config_entries = hass.config_entries.async_entries("zwave_js")
+    if not zwave_js_config_entries:
+        return False
+
+    migrated = any(
+        config_entry.data.get("migrated") for config_entry in zwave_js_config_entries
+    )
+    return migrated
+
+
+@callback
+def async_add_migration_entity_value(
+    hass: HomeAssistant,
+    entity_id: str,
+    entity_values: ZWaveDeviceEntityValues,
+) -> None:
+    """Add Z-Wave entity value for legacy Z-Wave migration."""
+    migration_handler: LegacyZWaveMigration = get_legacy_zwave_migration(hass)
+    migration_handler.add_entity_value(entity_id, entity_values)
+
+
+async def async_get_migration_data(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> dict[str, ZWaveMigrationData]:
+    """Return Z-Wave migration data."""
+    migration_handler = get_legacy_zwave_migration(hass)
+    return await migration_handler.get_data(config_entry)
+
+
+@singleton(LEGACY_ZWAVE_MIGRATION)
+@callback
+def get_legacy_zwave_migration(hass: HomeAssistant) -> LegacyZWaveMigration:
```

Comment:
We generally make these async and then include the loading of data. Is that not done here on purpose ?

Reply:
No, we should load too, so I'll change it.